sys/kern/subr_uio.c

```diff
@@ -72,6 +72,29 @@ SYSCTL_INT(_kern, KERN_IOV_MAX, iov_max, CTLFLAG_RD, SYSCTL_NULL_INT_PTR, UIO_MA
 static int uiomove_flags(void *cp, int n, struct uio *uio, bool nofault,
     bool preserve_tags);
 
+struct uio *
+alloc_uio(u_int iovcnt)
+{
+	struct uio *uio;
+	size_t iovlen;
+	size_t align = sizeof (void *);
+
+	iovlen = CHERI_REPRESENTABLE_LENGTH(iovcnt * sizeof (struct iovec));
+#ifdef __CHERI_PURE_CAPABILITY__
+	align = max(align, CHERI_REPRESENTABLE_ALIGNMENT(iovlen));
+#endif
+	/*
+	 * Can not set tight bounds on uio because free() requires a capability
+	 * with full bounds for the allocation.
+	 * XXX-AM: To avoid this, we must use two separate allocations.
+	 */
+	uio = malloc(iovlen + roundup2(sizeof (*uio), align), M_IOV, M_WAITOK);
+	uio->uio_iov = cheri_kern_setboundsexact(
+		(struct iovec *)(roundup2(uio + 1, align)), iovlen);
```

Comment:
I think you want to cast `uio + 1` to `uintptr_t`? Or does the builtin DTRT?

Reply:
The built-in should work on both integers and pointers, maintaining the tag.